pkgs/build-support/dart/pub2nix/pubspec-lock.nix

```diff
@@ -0,0 +1,117 @@
+{ lib
+, callPackage
+, fetchurl
+, fetchgit
+, runCommand
+}:
+
+{
+  # The source directory of the package.
+  src
+
+  # The package subdirectory within src.
+  # Useful if the package references sibling packages with relative paths.
+, packageRoot ? "."
+
+  # The pubspec.lock file, in attribute set form.
+, pubspecLock
+
+  # Hashes for Git dependencies.
+  # Pub does not record these itself, so they must be manually provided.
+, gitHashes ? { }
+
+  # Functions to generate SDK package sources.
+  # The function names should match the SDK names, and the package name is given as an argument.
+, sdkSourceBuilders ? { }
+
+  # Functions that create custom package source derivations.
+  #
+  # The function names should match the package names, and the package version,
+  # source, and source files are given in an attribute set argument.
+  #
+  # The passthru of the source derivation should be propagated.
+, customSourceBuilders ? { }
+}:
+
+let
+  dependencyVersions = builtins.mapAttrs (name: details: details.version) pubspecLock.packages;
+
+  dependencyTypes = {
+    "direct main" = "main";
+    "direct dev" = "dev";
+    "direct overridden" = "overridden";
+    "transitive" = "transitive";
+  };
+
+  dependencies = lib.foldlAttrs
+    (dependencies: name: details: dependencies // { ${dependencyTypes.${details.dependency}} = dependencies.${dependencyTypes.${details.dependency}} ++ [ name ]; })
+    (lib.genAttrs (builtins.attrValues dependencyTypes) (dependencyType: [ ]))
+    pubspecLock.packages;
+
+  mkHostedDependencySource = name: details:
+    let
+      archive = fetchurl {
```

Comment:
why not use fetchTarBall?

Reply:
added disclaimer why not

Reply:
The other reason is that the hash in the `pubspec.lock` is for the archive itself. There's no way to determine the recursive hash of the contents in advance.

Reply:
Ah I see. Is this correct then?
```
  # The hash in `pubspec.lock` is for the archive itself.
  # If we use `fetchTarball` instead of `fetchurl`, we'll need to input the hash for the files after extraction, which we don't have.
  # Additionally, `fetchTarball` fails with "tarball contains an unexpected number of top-level files".
  # https://discourse.nixos.org/t/fetchtarball-with-multiple-top-level-directories-fails/20556
```

Reply:
Looks good to me